src/Common/src/CoreLib/System/SpanHelpers.cs

```diff
@@ -461,43 +461,78 @@ public static unsafe void ClearWithoutReferences(ref byte b, nuint byteLength)
+                Unsafe.Add(ref Unsafe.Add(ref ip, (IntPtr)pointerSizeLength), -6) = default(IntPtr);
+                Unsafe.Add(ref Unsafe.Add(ref ip, (IntPtr)pointerSizeLength), -7) = default(IntPtr);
+                Unsafe.Add(ref Unsafe.Add(ref ip, (IntPtr)pointerSizeLength), -8) = default(IntPtr);
             }
-            if ((n = i + 4) <= (pointerSizeLength))
+
+            Debug.Assert(pointerSizeLength <= 7);
+
+            // The logic below works by trying to minimize the number of branches taken for any
+            // given range of lengths. For example, the lengths [ 4 .. 7 ] are handled by a single
+            // branch, [ 2 .. 3 ] are handled by a single branch, and [ 1 ] is handled by a single
+            // branch.
+            // 
+            // We can write both forward and backward as a perf improvement. For example,
+            // the lengths [ 4 .. 7 ] can be handled by zeroing out the first four natural
+            // words and the last 3 natural words. In the best case (length = 7), there are
+            // no overlapping writes. In the worst case (length = 4), there are three
+            // overlapping writes near the middle of the buffer. In perf testing, the
+            // penalty for performing duplicate writes is less expensive than the penalty
+            // for complex branching.
+
+            if (pointerSizeLength >= 4)
             {
-                Unsafe.AddByteOffset<IntPtr>(ref ip, (i + 0) * (nuint)sizeof(IntPtr)) = default(IntPtr);
-                Unsafe.AddByteOffset<IntPtr>(ref ip, (i + 1) * (nuint)sizeof(IntPtr)) = default(IntPtr);
-                Unsafe.AddByteOffset<IntPtr>(ref ip, (i + 2) * (nuint)sizeof(IntPtr)) = default(IntPtr);
-                Unsafe.AddByteOffset<IntPtr>(ref ip, (i + 3) * (nuint)sizeof(IntPtr)) = default(IntPtr);
-                i = n;
+                goto Write4To7;
             }
-            if ((n = i + 2) <= (pointerSizeLength))
+            else if (pointerSizeLength >= 2)
             {
-                Unsafe.AddByteOffset<IntPtr>(ref ip, (i + 0) * (nuint)sizeof(IntPtr)) = default(IntPtr);
-                Unsafe.AddByteOffset<IntPtr>(ref ip, (i + 1) * (nuint)sizeof(IntPtr)) = default(IntPtr);
-                i = n;
+                goto Write2To3;
             }
-            if ((i + 1) <= (pointerSizeLength))
+            else if (pointerSizeLength > 0)
             {
-                Unsafe.AddByteOffset<IntPtr>(ref ip, (i + 0) * (nuint)sizeof(IntPtr)) = default(IntPtr);
+                goto Write1;
             }
+            else
+            {
+                return; // nothing to write
+            }
+
+        Write4To7:
+            Debug.Assert(pointerSizeLength >= 4);
+
+            // Write first four and last three.
+            Unsafe.Add(ref ip, 2) = default(IntPtr);
+            Unsafe.Add(ref ip, 3) = default(IntPtr);
+            Unsafe.Add(ref ip, 4) = default(IntPtr);
```

Comment:
This line shouldn't be here. Will fix.

Reply:
Can you explain why it shouldn't be there?

Reply:
The element at index 4 is the *fifth* element in the span, which is a potential buffer overrun. I've checked in a fix directly to https://github.com/dotnet/corefx/commit/ca6964581f2067372741c8110db355de399e3cbc because this is already on its way to the coreclr train.

Reply:
cc @jkotas, @anipik - regarding mirroring

Reply:
I see that SpanHelpers.cs is included in portable System.Memory. It does not work - files that use nuint can be used for fast Span only. They cannot be used for portable System.Memory.  Portable System.Memory is not compiled bitness specific. Could you please fix that?

Reply:
Opened https://github.com/dotnet/corefx/issues/27774

Reply:
False alarm. This was done in CoreLib specific copy of SpanHelpers.cs. Please checkin the changes into CoreLib specific files through CoreCLR next time.

Reply:
We should consider renaming the file (SpanHelpers.Portable.cs and SpanHelpers.Fast.cs) to avoid the confusion. Thoughts?

Reply:
@ahsonkhan Not a bad idea. What repos consume this file? Just coreclr? coreclr and corert?

Reply:
Coreclr and Corert.